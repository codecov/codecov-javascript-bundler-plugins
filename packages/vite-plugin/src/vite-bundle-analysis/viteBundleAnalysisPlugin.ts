import path from "node:path";
import {
  normalizePath,
  type Asset,
  type Chunk,
  type Module,
  type BundleAnalysisUploadPlugin,
  red,
} from "@codecov/bundler-plugin-core";

const PLUGIN_NAME = "codecov-vite-bundle-analysis-plugin";

export const viteBundleAnalysisPlugin: BundleAnalysisUploadPlugin = ({
  output,
  userOptions,
}) => ({
  version: "1",
  name: PLUGIN_NAME,
  pluginVersion: "1.0.0",
  vite: {
    generateBundle(this, options, bundle) {
      // don't need to do anything if the bundle name is not present or empty
      if (!userOptions.bundleName || userOptions.bundleName === "") {
        red("Bundle name is not present or empty. Skipping upload.");
        return;
      }

      // append bundle output format to bundle name
      output.bundleName = `${userOptions.bundleName}-${options.format}`;

      // add in bundle name if present
      if (options.name && options.name !== "") {
        output.bundleName = `${userOptions.bundleName}-${options.name}`;
      }

      const cwd = process.cwd();
      const assets: Asset[] = [];
      const chunks: Chunk[] = [];
      const moduleByFileName = new Map<string, Module>();
      const items = Object.values(bundle);
      const customOptions = {
        moduleOriginalSize: false,
        ...options,
      };

      let assetFormatString = "";
      if (typeof customOptions.assetFileNames === "string") {
        assetFormatString = customOptions.assetFileNames;
      }

      let chunkFormatString = "";
      if (typeof customOptions.chunkFileNames === "string") {
        chunkFormatString = customOptions.chunkFileNames;
      }

      let counter = 0;
      for (const item of items) {
        if (item?.type === "asset") {
          if (typeof item.source === "string") {
            const fileName = item?.fileName ?? "";
            const size = Buffer.from(item.source).byteLength;

            assets.push({
              name: fileName,
              size: size,
              normalized: normalizePath(fileName, assetFormatString),
            });
          } else {
            const fileName = item?.fileName ?? "";
            const size = item?.source.byteLength;

            assets.push({
              name: fileName,
              size: size,
              normalized: normalizePath(fileName, assetFormatString),
            });
          }
        }

        if (item?.type === "chunk") {
          const chunkId = item?.name ?? "";
          const fileName = item?.fileName ?? "";
          const moduleEntries = Object.entries(item?.modules ?? {});
          const size = Buffer.from(item?.code).byteLength;
          const uniqueId = `${counter}-${chunkId}`;

          assets.push({
            name: fileName,
            size: size,
            normalized: normalizePath(fileName, chunkFormatString),
          });

          chunks.push({
            id: chunkId,
            uniqueId: uniqueId,
            entry: item?.isEntry,
            initial: item?.isDynamicEntry,
            files: [fileName],
            names: [item?.name],
          });

          for (const [modulePath, moduleInfo] of moduleEntries) {
            const normalizedModulePath = modulePath.replace("\u0000", "");
            const relativeModulePath = path.relative(cwd, normalizedModulePath);

            const relativeModulePathWithPrefix = relativeModulePath.match(
              /^\.\./,
            )
              ? relativeModulePath
              : `.${path.sep}${relativeModulePath}`;

            // try to grab module already set in map
            const moduleEntry = moduleByFileName.get(
              relativeModulePathWithPrefix,
            );

            // if the modules exists append chunk ids to the grabbed module
            // else create a new module and create a new entry in the map
            if (moduleEntry) {
              moduleEntry.chunkUniqueIds.push(uniqueId);
            } else {
              const size = customOptions.moduleOriginalSize
                ? moduleInfo.originalLength
                : moduleInfo.renderedLength;

              const module: Module = {
                name: relativeModulePathWithPrefix,
                size: size,
                chunkUniqueIds: [uniqueId],
              };

              moduleByFileName.set(relativeModulePathWithPrefix, module);
            }
          }
          counter += 1;
        }
      }

      // grab the modules from the map and convert it to an array
      const modules = Array.from(moduleByFileName.values());

      output.bundler = {
        name: "rollup",
        version: this.meta.rollupVersion,
      };
      output.assets = assets;
      output.chunks = chunks;
      output.modules = modules;

      // only output file if running dry run
      if (userOptions?.dryRun) {
        this.emitFile({
          type: "asset",
          fileName: "codecov-bundle-stats.json",
          source: JSON.stringify(output),
        });
      }
    },
  },
});
