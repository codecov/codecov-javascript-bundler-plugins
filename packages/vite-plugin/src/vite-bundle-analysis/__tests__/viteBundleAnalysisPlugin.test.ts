import { viteBundleAnalysisPlugin } from "../viteBundleAnalysisPlugin";

describe("viteBundleAnalysisPlugin", () => {
  describe("when called", () => {
    it("returns a plugin object", () => {
      const plugin = viteBundleAnalysisPlugin({
        output: {},
        userOptions: {},
      });

      expect(plugin.version).toEqual("1");
      expect(plugin.name).toEqual("codecov-vite-bundle-analysis-plugin");
      expect(plugin.pluginVersion).toEqual("1.0.0");
      expect(plugin.vite?.generateBundle).toEqual(expect.any(Function));
    });
  });
});
