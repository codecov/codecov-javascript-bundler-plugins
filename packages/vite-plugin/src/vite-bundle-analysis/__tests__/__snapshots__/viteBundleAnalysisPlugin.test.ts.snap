// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viteBundleAnalysisPlugin > when called > returns a plugin object 1`] = `
{
  "buildEnd": [Function],
  "buildStart": [Function],
  "name": "@codecov/vite-plugin",
  "pluginVersion": "1.2.1",
  "version": "2",
  "vite": {
    "generateBundle": [Function],
  },
  "writeBundle": [Function],
}
`;
